### PR TITLE
Create builder for BDDs and ZDDs

### LINKED ISSUE #325: Further simplify manual BDD/ZDD creation
With the `node_writer` we still require the user to use all of the functions in *data.h*, to understand how a file is laid out, and to manage information themselves. Instead, we can have a new class `builder<dd_policy>` that hides the `node_file`, the `node_writer` and everything in *data.h* away.
```cpp
/////////////////////////////////////////////////////////////////////////////
/// \brief Add a terminal node with a given value.
/////////////////////////////////////////////////////////////////////////////
ptr_t
builder<typename dd_policy>::add_node(bool sink_value);

/////////////////////////////////////////////////////////////////////////////
/// \brief Add an internal node with a given label and children.
/////////////////////////////////////////////////////////////////////////////
ptr_t
builder<typename dd_policy>::add_node(label_t label, ptr_t low, ptr_t high);

/////////////////////////////////////////////////////////////////////////////
/// \brief Obtain the final decision diagram.
/////////////////////////////////////////////////////////////////////////////
typename dd_policy::reduced_t
builder<typename dd_policy>::create();

/////////////////////////////////////////////////////////////////////////////
/// \brief Reset to create a new decision diagram.
/////////////////////////////////////////////////////////////////////////////
void
builder<typename dd_policy>::reset();
```
This further abstracts away the pesky implementation details of BDDs in Adiar. At this point the user only needs to create nodes bottom-up and remember to store the returned `ptr_t` for use later. Both of those things should seem (with a little documentation) intuitive.

- [x] Create `builder<dd_policy>` and then *typedef* the `bdd_builder` and `zdd_builder` to hide away the policy in question.

  Some extra desired features would be.
  - [x] *add_node* with a sink does nothing, but remembers that it was called. If *create* is called immediately afterwards, then construct a sink.
   - [x] Replace all `adiar_assert` statements with throwing an exception.
   - [x] Write a proper set of unit tests for the builder. Assuming it uses exceptions then these can also be tested for (yay!)
   - [x] Apply *Reduction Rule 1* at the very beginning of `make_node`.
  
  We can further improve the builder by making the *add_node* functions not return a `ptr_t` but a wrapper class `ptr<dd_policy>` instead.
  - [x] To identify whether there is only a single root, we need to check that every returned non-sink node has been referenced by some later call to *add_node*. To do this, add a `std::shared_ptr<ptr_t>` to both the `builder<dd_policy>` and the `ptr<dd_policy>` class. Furthermore, add a boolean `unreferenced` (initially `true`) and is set to `false` in *add_node*.
  
      On destruct `~ptr<dd_policy>() { ... }` check if it is *unreferenced* and set the value of the `std::shared_ptr<ptr_t>` to the wrapped `ptr_t`. If `std::shared_ptr<ptr_t>` is **not** `NIL` then we have two roots.
  - [x] We can identify cross-polution, i.e. using pointers from one builder in another, by checking whether the `std::shared_ptr<ptr_t>` actually points to the same address in memory.
- [x] Clean up the *node_writer* by moving the relevant responsibilities into *builder<dd_policy>*. Move as much of the user validation logic from the *push* / *<<* function out of the *node_writer* and into the builder. That is, keep everything related to *canonicity* and cuts and so on inside of the *node_writer*.

### PR DESCRIPTION
**Pull request by AnnaBlume99**
Closes #325

### CODE REVIEW

File: src/adiar/builder.h [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+      if(ptr == node.uid) {
+        nw.push(node);
+        unref_nodes++;
+      }
+
+      //Update count of unreferenced nodes
+      bool low_unref = *(low.unreferenced.get());
+      if(low_unref && !is_sink(low.ptr)) {
+        std::shared_ptr<bool> low_p = std::make_shared<bool>(false);
+        low.unreferenced.swap(low_p);
```
Comment (SSoelvsten) [inline src/adiar/builder.h:109]: This breaks, since you replace the pointer-content of the `builder_ptr` to point to something completely different. So, if you have two copies of the *low* one of them is updated to point to a new *false* value, while the other one still points to the old *true* value. Or, possibly the *swap* value makes it all work anyways, but there is a much simpler solution: update the value that is being pointed to:
```cpp
*low.unreferenced = false;
```

File: src/adiar/builder.h [hunk 2]
```diff
@@ -0,0 +1,184 @@
...
+        unref_nodes--;
+      }
+      bool high_unref = *(high.unreferenced.get());
+      if(high_unref && !is_sink(high.ptr)) {
+        std::shared_ptr<bool> high_p = std::make_shared<bool>(false);
+        high.unreferenced.swap(high_p);
```
Comment (SSoelvsten) [inline src/adiar/builder.h:115]: *high.unreferenced = false;

File: src/adiar/builder.h (hunk 1 above)
Comment (SSoelvsten) [reply src/adiar/builder.h:109]: In fact, if you change line 106 to be
```cpp
bool& low_unref = *(low.unreferenced.get());
```
then you should be able to just do
```cpp
low_unref = false;
```

File: src/adiar/builder.h [hunk 3]
```diff
@@ -0,0 +1,184 @@
...
+        unref_nodes--;
+      }
+
+      return make_ptr(ptr);
+    }
```
Comment (SSoelvsten) [inline src/adiar/builder.h:120]: I believe a lot of this would be easier to understand, if the *reduction rule your removes node* is an early-termination edge-case.
```cpp
node_t node = create_node(label, current_id, low.ptr, high.ptr);

ptr_t ptr = dd_policy::reduction_rule(node);
if (ptr == low.ptr) { return low; }
if (ptr == high.ptr) { return high; }

nw.push(node);
unref_nodes++;
current_id--;

...
```
Especially this way, one may mark a node as *unreferenced* even if it is yet not referenced by anything (because the referencing node does not exist). In fact, this way we would allow this multi-rooted BDD
```
        *
        ||
    1   2
   / \ /
  3   4
 / \ / \
 F T T F
```
This edge-case behaviour should be captured by a few extra unit tests.

File: test/adiar/test_builder.cpp [hunk 4]
```diff
@@ -0,0 +1,370 @@
...
+    });
+
+    it("throws an exception if pointers are used from a different builder", [&]() {
+      bdd_builder b1;
+      bdd_builder b2;
+
+      builder_ptr p = b1.add_node(true);
+
+      AssertThrows(std::invalid_argument, b2.add_node(0,false,p));
+    });
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:107]: Maybe a second instance of this test with a non-sink pointer would be nice.

File: test/adiar/test_builder.cpp [hunk 5]
```diff
@@ -0,0 +1,370 @@
...
+      AssertThrows(std::invalid_argument, b.add_node(0,p,true));
+    });
+
+    it("throws an exception when label_of(high) >= label", [&]() {
+      bdd_builder b;
+
+      bdd_ptr p = b.add_node(0,false,true);
+
+      AssertThrows(std::invalid_argument, b.add_node(0,false,p));
+    });
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:137]: Maybe a second copy of the same unit tests with a non-zero label?

File: test/adiar/test_builder.cpp [hunk 6]
```diff
@@ -0,0 +1,370 @@
...
+
+    it("throws an exception when there is more than one root", [&]() {
+      bdd_builder b;
+
+      b.add_node(0,false,true);
+
+      b.add_node(0,true,false);
+
+      AssertThrows(std::domain_error, b.create());
+    });
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:309]: A second large unit test like this (also with non-zero labels) would be great.

File: test/adiar/test_builder.cpp [hunk 7]
```diff
@@ -0,0 +1,370 @@
...
+
+    describe("bdd_builder", [&]() {
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:311]: Missing unit test, that the reduction rule does not mess up the multi-root checking.

File: test/adiar/test_builder.cpp [hunk 8]
```diff
@@ -0,0 +1,370 @@
...
+        level_info_test_stream<node_t> out_meta(out);
+
+        AssertThat(out_meta.can_pull(), Is().True());
+        AssertThat(out_meta.pull(), Is().EqualTo(create_level_info(1,1u)));
+
+        AssertThat(out_meta.can_pull(), Is().False());
+      });
+    });
+
+    describe("zdd_builder", [&]() {
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:340]: Missing unit test, that the reduction rule does not mess up the multi-root checking.

File: test/adiar/test_builder.cpp [hunk 9]
```diff
@@ -0,0 +1,370 @@
...
+
+        AssertThat(out_meta.can_pull(), Is().True());
+        AssertThat(out_meta.pull(), Is().EqualTo(create_level_info(1,1u)));
+
+        AssertThat(out_meta.can_pull(), Is().False());
+      });
+    });
+
+  });
+});
```
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:370]: Missing new line at end of file.

File: test/adiar/test_builder.cpp (hunk 6 above)
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:309]: A unit test for this check, that also uses a copy-constructor would be great.
```cpp
bdd_ptr p1 = b.add_node(...);
bdd_ptr p2 = p1;
bdd_ptr p3 = b.add_node(..., p2, ...);
```

File: test/adiar/test_builder.cpp (hunk 7 above)
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:311]: Missing unit test for the reduction rule together with the copy-constructor.
```cpp
bdd_ptr p1 = b.add_node(...);
bdd_ptr p2 = p1;
bdd_ptr p3 = b.add_node(..., p1, p2);
```

File: test/adiar/test_builder.cpp (hunk 4 above)
Comment (SSoelvsten) [inline test/adiar/test_builder.cpp:107]: A test to catch the *low* pointer being the culprit is missing.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
Good first draft. 🙂 There is a bit to be done, but pretty much all of it are small things.

File: src/adiar/builder.h [hunk 10]
```diff
@@ -0,0 +1,222 @@
...
+  struct builder_shared
+  {
+
+  };
+
+  ///////////////////////////////////////////////////////////////////////////////
+  /// \brief A builder for decision diagrams
+  ///
+  /// \sa builder
+  ///////////////////////////////////////////////////////////////////////////////
```
Comment (SSoelvsten) [inline src/adiar/builder.h:21]: No need for two comments of the same thing.

File: src/adiar/builder.h [hunk 11]
```diff
@@ -0,0 +1,184 @@
...
+    std::shared_ptr<bool> unreferenced = std::make_shared<bool>(true);
+
+  public:
+    builder_ptr(ptr_t p, std::shared_ptr<ptr_t> sp) : ptr(p), shared_ptr(sp) { }
+
+    ~builder_ptr() { }
+  };
+
+  typedef builder_ptr<bdd_policy> bdd_ptr;
+  typedef builder_ptr<zdd_policy> zdd_ptr;
```
[the remark this answers is not in the document]
Comment (SSoelvsten) [reply src/adiar/builder.h:58]: The comment needs to be for each independently. Right now it only applies to `bdd_ptr`.

File: src/adiar/builder.h [hunk 12]
```diff
@@ -0,0 +1,184 @@
...
+    }
+
+  private:
+    builder_ptr<dd_policy> make_ptr(ptr_t p) {
+      return builder_ptr<dd_policy>(p, shared_ptr);
+    }
+  };
+
+  typedef builder<bdd_policy> bdd_builder;
+  typedef builder<zdd_policy> zdd_builder;
```
[the remark this answers is not in the document]
Comment (SSoelvsten) [reply src/adiar/builder.h:218]: The comment needs to be for each independently. Right now it only applies to `bdd_builder`.

File: src/adiar/file_writer.h [hunk 13]
```diff
@@ -506,8 +491,11 @@ namespace adiar {
     /// \brief Attach to a file
     ////////////////////////////////////////////////////////////////////////////
     void attach(const node_file &f) {
-      // TODO: set _latest_node etc. when opening file
+      // TODO: set cuts, etc. when opening file
```
Comment (SSoelvsten) [inline src/adiar/file_writer.h:494]: It's also important, that the cuts are correct. So, this reset should also set those four values back to their initial values.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
Two things:

- Nitpicks about the documentation comments
- It is also important, that it generates the correct cut sizes. That is, the `.detach()` on *node_writer* should also reset the cut bounds.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
The unit tests are missing tests for the cuts. Here is code from *internal/test_reduce.cpp* that can be repurposed.
```cpp
        AssertThat(out->max_2level_cut[cut_type::INTERNAL], Is().GreaterThanOrEqualTo(2u));
        AssertThat(out->max_2level_cut[cut_type::INTERNAL], Is().LessThanOrEqualTo(4u));
        AssertThat(out->max_2level_cut[cut_type::INTERNAL_FALSE], Is().GreaterThanOrEqualTo(3u));
        AssertThat(out->max_2level_cut[cut_type::INTERNAL_FALSE], Is().LessThanOrEqualTo(5u));
        AssertThat(out->max_2level_cut[cut_type::INTERNAL_TRUE], Is().GreaterThanOrEqualTo(3u));
        AssertThat(out->max_2level_cut[cut_type::INTERNAL_TRUE], Is().LessThanOrEqualTo(5u));
        AssertThat(out->max_2level_cut[cut_type::ALL], Is().GreaterThanOrEqualTo(4u));
        AssertThat(out->max_2level_cut[cut_type::ALL], Is().LessThanOrEqualTo(5u));
```
Of course, both the 1-level and 2-level cases should be added. The upper bound comes directly from the minimum of
1. *N+1* where *N* is the number of nodes
2. *A* where *A* is the number of arcs of that type

And for 2-level cuts also:
3. A 3/2 factor on the 1-level cut.

The lower bound should be the actual value of the diagram.

**Review by SSoelvsten — CHANGES_REQUESTED** [review]
Found a minor problem with the `builder_ptr` default constructor. Other than that it is ready for a merge.

**Review by SSoelvsten — APPROVED** [review]
